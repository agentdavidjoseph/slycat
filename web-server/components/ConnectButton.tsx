import * as React from 'react';
import client from "../js/slycat-web-client";

export interface ConnectButtonProps { 
  loadingData: boolean
  hostname: string
  username: string
  password: string
  callBack: Function
  sessionExists: boolean
  text?:string
}

export interface ConnectButtonState {
  loadingData: boolean
  sessionExists: boolean
  text: string
}

export default class RemoteFileBrowser extends React.Component<ConnectButtonProps, ConnectButtonState> {
  public constructor(props:ConnectButtonProps) {
    super(props)
    this.state = {
      text: this.props.text?this.props.text:"Connect",
      loadingData: props.loadingData,
      sessionExists: props.sessionExists
    }
  }

  /**
   * function used to test if we have an ssh connection to the hostname
   * @param {hostname}
   * @memberof SlycatRemoteControls
   */
  checkRemoteStatus = async (hostname:string) => {
    return client.get_remotes_fetch(hostname)
      .then((json:any) => {
        console.log(`calling the callback ${json.status}`)
        this.setState({
          sessionExists:json.status,
          loadingData:false
        }, () => {
          console.log(`calling the callback ${this.state.sessionExists}`)
          this.props.callBack(this.state.sessionExists, this.state.loadingData)
        });
    });
  };

  /**
   * establishes a new ssh connection given the
   * hostname, username, password
   *
   * @memberof SlycatRemoteControls
   */
  connect = async () => {
    this.setState({loadingData:true})
    this.props.callBack(this.state.sessionExists, true);
    client.post_remotes_fetch({
      parameters: {
        hostname: this.props.hostname,
        username: this.props.username,
        password: this.props.password,
      }
    }).then(() => {
      this.checkRemoteStatus(this.props.hostname);
      console.log("Remote session created.");
    }).catch((errorResponse:any) => {
      if (errorResponse.status == 403){
        alert(`${errorResponse.statusText} \n\n-Make sure your username and password are entered correctly. 
        \n-Note you also may have tried to many times with bad credentials and been suspended for the next few minutes`)
      } else if (errorResponse.status == 401){
        alert(`${errorResponse.statusText} \n\n-Make sure the Hostname is entered correctly.`)
      } else {
        alert(`${errorResponse.statusText}`)
      }
      this.setState({loadingData:false}, () => {
        this.props.callBack(this.state.sessionExists, this.state.loadingData);
      })
    });
  };

  public render() {
    return (
      <button disabled={this.state.loadingData} type='button' className='btn btn-primary float-right' onClick={this.connect}>
        {this.state.loadingData?<span className="spinner-border spinner-border-sm" role="status" aria-hidden="true"></span>:null}
        {this.state.loadingData?"Loading...":this.state.text}
      </button>
    );
  }
}
