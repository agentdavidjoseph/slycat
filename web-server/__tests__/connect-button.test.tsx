
import React from 'react';
import { mount } from "enzyme";
import ConnectButton, {ConnectButtonProps, ConnectButtonState} from 'components/ConnectButton';
interface GlobalAny extends NodeJS.Global {
  fetch: any
}
const globalAny:GlobalAny = global;
describe('when loading a nav bar',() =>{ 
  test('we have expected props on initial load', async () => {
    const properties: ConnectButtonProps = {
      text: 'buttonText',
      callBack: jest.fn(),
      sessionExists:false,
      loadingData:false,
      hostname:'testHost',
      username:'user',
      password:'password123'
    };
    const render = await mount(
      <ConnectButton
        {...properties}
      />
        );
    const props = render.props() as ConnectButtonProps;
    expect(props).toMatchSnapshot();
  });

  test('we have proper state', async () => {
    const properties: ConnectButtonProps = {
      text: 'buttonText',
      callBack: jest.fn(),
      sessionExists:false,
      loadingData:false,
      hostname:'testHost',
      username:'user',
      password:'password123'
    };
    const render = await mount(
      <ConnectButton
        {...properties}
      />
        );
    const state = render.state() as ConnectButtonState;
    expect(state).toMatchSnapshot();
  });

  test('we should have a full initialization', async () => {
    const properties: ConnectButtonProps = {
      text: 'buttonText',
      callBack: jest.fn(),
      sessionExists:false,
      loadingData:false,
      hostname:'testHost',
      username:'user',
      password:'password123'
    };
    const render = await mount(
      <ConnectButton
        {...properties}
      />
        );
    expect(render.instance()).toMatchSnapshot();
  });

  test('we be able to check hostname status', async () => {
    const properties: ConnectButtonProps = {
      text: 'buttonText',
      callBack: jest.fn(),
      sessionExists:false,
      loadingData:false,
      hostname:'testHost',
      username:'user',
      password:'password123'
    };
    globalAny.fetch.mockResponseOnce(JSON.stringify({ status: true }));
    const render = await mount(
      <ConnectButton
        {...properties}
      />
    );
    //make our webservice call
    await (render.instance() as any).checkRemoteStatus();
    const newState =render.state() as ConnectButtonState;
    expect(newState.sessionExists).toBe(true);
  });

  test('we be able to call connect', async () => {
    const properties: ConnectButtonProps = {
      text: 'buttonText',
      callBack: jest.fn(),
      sessionExists:false,
      loadingData:false,
      hostname:'testHost',
      username:'user',
      password:'password123'
    };
    globalAny.fetch.mockResponse(JSON.stringify({ status: true }));
    const render = await mount(
      <ConnectButton
        {...properties}
      />
    );
    //make our webservice call
    const instance:any = render.instance() as any;
    jest.spyOn(instance, 'connect');
    await instance.connect();
    expect(instance.connect).toHaveBeenCalled();
  });
});

export default undefined