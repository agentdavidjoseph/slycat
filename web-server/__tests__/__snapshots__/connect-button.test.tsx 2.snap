// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when loading a nav bar we have expected props on initial load 1`] = `
Object {
  "callBack": [MockFunction],
  "hostname": "testHost",
  "loadingData": false,
  "password": "password123",
  "sessionExists": false,
  "text": "buttonText",
  "username": "user",
}
`;

exports[`when loading a nav bar we have proper state 1`] = `
Object {
  "loadingData": false,
  "sessionExists": false,
  "text": "Connect",
}
`;

exports[`when loading a nav bar we should have a full initialization 1`] = `
<ConnectButton
  callBack={[MockFunction]}
  hostname="testHost"
  loadingData={false}
  password="password123"
  sessionExists={false}
  text="buttonText"
  username="user"
>
  <button
    className="btn btn-primary float-right"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    buttonText
  </button>
</ConnectButton>
`;
